<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>A flexible multi-objective optimization framework for transferring features across single-cell and spatial omics • DOTr</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="A flexible multi-objective optimization framework for transferring features across single-cell and spatial omics">
<meta name="description" content="DOT is a method for transferring cell features from a reference single-cell RNA-seq data to spots/cells in spatial omics. It operates by optimizing a combination of multiple objectives using a Frank-Wolfe algorithm to produce a high quality transfer. Apart from transferring cell types/states to spatial omics, DOT can be used for transferring other relevant categorical or continuous phenotypes from one set of omics to another, such as estimating the expression of missinng genes or transferring transcription factor/pathway activities.">
<meta property="og:description" content="DOT is a method for transferring cell features from a reference single-cell RNA-seq data to spots/cells in spatial omics. It operates by optimizing a combination of multiple objectives using a Frank-Wolfe algorithm to produce a high quality transfer. Apart from transferring cell types/states to spatial omics, DOT can be used for transferring other relevant categorical or continuous phenotypes from one set of omics to another, such as estimating the expression of missinng genes or transferring transcription factor/pathway activities.">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">DOTr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/general.html">DOT: A flexible multi-objective optimization framework for transferring features across single-cell and spatial omics</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="dot-flexible-feature-transfer-to-spatial-omics-">DOT: Flexible Feature Transfer to Spatial Omics <img src="reference/figures/DOT_icon.png" align="right" height="139"><a class="anchor" aria-label="anchor" href="#dot-flexible-feature-transfer-to-spatial-omics-"></a>
</h1></div>
<!-- badges: start -->

<!-- badges: end -->
<p>## Overview</p>
<p><code>DOT</code> is a method for transferring cell features from a reference single-cell RNA-seq data to spots/cells in spatial omics. It operates by optimizing a combination of multiple objectives using a Frank-Wolfe algorithm to produce a high quality transfer. Apart from transferring cell types/states to spatial omics, <code>DOT</code> can be used for transferring other relevant categorical or continuous features from one set of omics to another, such as estimating the expression of missinng genes or transferring transcription factor/pathway activities.</p>
<p align="center" width="100%">
<img src="reference/figures/overview.png" align="center" width="65%"></p>
<p>For more information about how this package has been used with real data and expected outputs, please check the following link:</p>
<ul>
<li><a href="https://saezlab.github.io/DOT/articles/general.html">DOT’s general usage</a></li>
</ul>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p><code>DOT</code> is available under the R package <code>DOTr</code> which you can install from <a href="https://github.com/" class="external-link">GitHub</a> with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"saezlab/DOT"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="dependencies">Dependencies<a class="anchor" aria-label="anchor" href="#dependencies"></a>
</h2>
<ul>
<li>R (&gt;= 4.0)</li>
<li>R packages: fields, ggplot2, Matrix, methods, Seurat, stats, stringr</li>
</ul>
<p>For optimal performance on moderately sized instances, we recommend at least 4 GB of RAM. For large reference scRNA-seq data or very large spatial instances higher memory may be required.</p>
<p>Installation takes less than five minutes. The sample dataset provided can be run in less than a minute on a “normal” desktop computer. DOT takes approximately 7 minutes to process a MERFISH MOp dataset with approximately 250 genes, 100 cell types and 4,000 spots.</p>
<p>Operating system tested on: macOS Monterey 12.4</p>
</div>
<div class="section level2">
<h2 id="python-version">Python version<a class="anchor" aria-label="anchor" href="#python-version"></a>
</h2>
<p>Erick Armignol (<a href="https://github.com/earmingol" class="external-link uri">https://github.com/earmingol</a>) has implemented a native Python version of DOT, supporting also deployment on GPUs.</p>
<p>Check out <a href="https://github.com/earmingol/DOTpy" class="external-link uri">https://github.com/earmingol/DOTpy</a> for information on how to install it, examples of use and more practical tips.</p>
<p>Thank you Erick!</p>
</div>
<div class="section level2">
<h2 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h2>
<p>If you use <strong>DOT</strong> for your research please cite the <a href="https://doi.org/10.1038/s41467-024-48868-z" class="external-link">following article</a>:</p>
<blockquote>
<p>Rahimi, A., Vale-Silva, L.A., Fälth Savitski, M. et al. DOT: a flexible multi-objective optimization framework for transferring features across single-cell and spatial omics. Nat Commun 15, 4994 (2024). <a href="https://doi.org/10.1038/s41467-024-48868-z" class="external-link uri">https://doi.org/10.1038/s41467-024-48868-z</a>.</p>
</blockquote>
</div>
</div>
  </main><aside class="col-md-3"><div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small>GPL (&gt;= 3)</small></li>
</ul>
</div>

<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing DOTr</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Arezou Rahimi <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0003-0285-8089" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a>  </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/saezlab/DOT/actions/workflows/R-CMD-check.yaml" class="external-link"><img src="https://github.com/saezlab/DOT/actions/workflows/R-CMD-check.yaml/badge.svg" alt="R-CMD-check"></a></li>
<li><a href="https://github.com/saezlab/DOT/issues" class="external-link"><img src="https://img.shields.io/github/issues/saezlab/DOT" alt="GitHub issues"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Arezou Rahimi.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
